# `Python` - Lesson 1

In this first lesson we will make a brief journey through the main feature of `python` language focusing on few specific topic that will be useful for the rest of the course.


## What is `python`

`Python` is a so called *interpreted* language, you write some code (a sequence of instructions or commands), this is read and executed. This is different from other programming languages like `C` or `C++` which compile code into a language that the computer can understand directly (*machine language*). As a result, `python` is essentially an *interactive* programming language, you can program and see the results almost at the same time.

![Interpreted vs compiled language](index.png)

`Python`, as basically all programs, comes in different version and flavours, the latest  is `3.8` (and it is continously evolving). 
**We will go for `python 3.7` !** because there are no critical difference with respect to the latest and because it is what you have installed in you computers...

Any Python interpreter, available at http://www.python.org, comes with a standard set of *packages*, but if you want more functionality, you can download more of them (there are zillions of packages out there). 

Some examples are:

* `numpy` - which provides matrix algebra functionality;
* `scipy` - which provides a whole series of scientific computing functions;
* `pandas` - which provides tools for manipulating time series or dataset in general;
* `matplotlib` - for plotting graphs;
* `jupyter` - for notebooks like this one;
* ...and many more.

### How we will use it

In the rest of the course you will be asked to use `Anaconda python` (https://www.anaconda.com).

`Anaconda` is a free and open-source distribution of the `python`programming languages for scientific computing, that aims to simplify package management. The distribution includes data-science packages suitable for any platform. Its graphical interface allows to easily install new modules and includes an IDE called `PyCharm`, an interactive shell to run quickly few lines of code and the `jupyter` application to write code and document it in a very easy way.

### Online courses

Python popularity is growing every day so it is very easy to find good (and free) online courses looking in Google. If you want to go deeper into the potentiality of this language I strongly suggest you to spend some time in watching one of them. The following one is quite basic but there are many others even on specific subjects you may like to focus on:

**MITx: 6.00.1x Introduction to Computer Science and Programming Using Python** 
https://courses.edx.org/courses/course-v1:MITx+6.00.1x+2T2017_2/course/ 

### Let's spend few minutes all together to setup `Anaconda`

## The language

In [1]:
# this is a comment 
# print is the simplest (and most used) keyword
print ("Hello world !") 

Hello world !


### Variables

Variables are used to store information to be referenced and manipulated in a computer program (e.g. a number, a string...).

![](var1.jpeg)

A `python` variable name must:
* begin with a letter (myphone) or underscore (\_myphone);
* other characters can be letters, numbers or more \_;
* variable names are case-sensitive so myphone and myPhone are two distinct variables;

**There are some reserved words which you cannot use as a variable name because Python uses them for other things (e.g. $\tt{print, type, for...}$)**.

To use GOOD variable names always choose meaningful names instead of short names (i.e. $\tt{numberOfCakes}$ is much better than simply $\tt{n}$), try to be consistent with your conventions (e.g. choose once and for all between $\tt{number\_of\_cakes}$ or $\tt{numberofcakes}$ or $\tt{numberOfCakes}$), usually begin a variable name with underscore (\_) only for a special case (will see later when this is usually done).

### Mathematical expressions

In [9]:
1 + 2

3

In [10]:
40 - 5

35

In [11]:
x * 20 # remember that we set x equal to 9

180

In [12]:
x / 4

2.25

In [14]:
x // 4 # interger division - result will be truncated to the 
       # corresponding integer (no rounding)
       # 11 / 3 = 3.666666 -> 11 // 3 = 3

2

In [15]:
y = 3
x ** y # x to the power of y

729

In [16]:
3 * (x + y)

36

### Modules

Useful functions can be saved in libraries (called *modules*) so that thay can be re-used in different programs. The keyword *import* allows to load functions and data from other Python files (*modules*) and make them available in your program.
Python already comes with a lot of built-in modules for doing lots of different tasks (the so called *standard library*) but many more modules are available for download on the web and you can of course write your own !

![Python has many modules for download on the web...](python.png)

In [1]:
import math

**Once a module has been imported, the functions it contains can be accessed with a *dot* "module_name.function_name":**

In [19]:
math.log(3)

1.0986122886681098

In [20]:
math.exp(3)

20.085536923187668

Since we are lazy and we don't want to type `math.` every time we compute a logarithm or an exponential, so we can use the following syntax:

In [3]:
from math import log, exp
print (log(3))
print (exp(3))

1.0986122886681098
20.085536923187668


Putting together what we have learned so far we can try to save the result of an expression in a variable (or even in the same variable we started with).
As an example let's compute the interest rate $r$ that produces a return $R$ of about 11051.71 Euro when investing 10000 Euro for 2 years:

$R = N\mathrm{e}^{r\tau} \rightarrow r = \frac{1}{\tau} \mathrm{log}(\frac{R}{N})$

In [4]:
rate = (1/2)*log(11051.71/10000)
print (rate)

0.05000003706410832


### Indented blocks and the $\tt{if/else}$ statement

Unlike other languages which uses parenthesis to isolate blocks of code Python uses indentation. A first example of this is given by the `if/then` statements where based on some condition we can run different part of the code.

In [37]:
# remember x value is 13 now
if x == 1: 
    print ("This will not be printed") 
    # the block of code that is run if the first condition is met is indented
elif x == 15:
    print ("This will not be printed either")
    # again the block of code that is run here is indented to be "isolated" by the rest 
else:
    print ("This *will* be printed")

This *will* be printed


In [38]:
# if by mistake I forget to indent some block I get an error
if x == 1: 
print ("This will not be printed")
elif x == 15:
    print ("This will not be printed either")
else:
    print ("This *will* be printed")

IndentationError: expected an indented block (<ipython-input-38-4535a45a6419>, line 3)

### Loops

Another very import feature of a language is the loop that allows to repeat the same block of code many times. In Python loops can be coded with `for` or `while` keywords.

#### `for` Loop

In [2]:
# the range keyword returns a "list" of integers 
# starting from first param to second param with
# steps given by the last parameter
for i in range(30, 25, -1):
    print (i)

30
29
28
27
26


In [43]:
for i in range(10):
    if i == 5:
        continue # this line skips the current iteration, 
                 # 5 is actually missing from the output below
    print (i)

0
1
2
3
4
6
7
8
9


In [44]:
for i in (4, 6, 10, 20): # here we loop directly on a list of numbers
    print (i)

4
6
10
20


In [45]:
phrase = 'how to loop over a string'
for c in phrase:
    print (c)

h
o
w
 
t
o
 
l
o
o
p
 
o
v
e
r
 
a
 
s
t
r
i
n
g


#### `while` Loop

In a for loop we go through all the elements of a list of objects, the `while` statement instead repeats the same block of code until a condition is met.

In [46]:
x = 1
while x ** 2 < 50: # the following block of code is run if x squared is <50
    print (x)
    x += 1 # each time we increment by 1 the variable x

1
2
3
4
5
6
7


It is possible to exit prematurely from a while loop using the `break` keyword.

In [47]:
x = 1
while True: # True is always true :-) so this would run forever
    if (x ** 2 > 50): 
        break # this line exit from the while loop
    print (x)
    x += 1 

1
2
3
4
5
6
7


### Lists

A list in `python` is a container that is a *mutable*, ordered sequence of elements. Each element or value that is inside of a list is called an item. Each item can be accessed using squared brackets (list indexing is zero-based). It is considered mutable since you can add, remove or update the items in the list. Ordered instead means that items are kept in the same order they have been added to the list.

In [48]:
mylist = [21, 32, 15]
mylist

[21, 32, 15]

In [51]:
len(mylist) # len() returns the number of items in a list

3

In [52]:
mylist[0] # accessing an item by index,remember the first item is element 0

21

In [53]:
mylist[1] = 74 # we can change list items since it's *mutable*
mylist

[21, 74, 15]

In [54]:
mylist[3] # error ! it doesn't exists, the list has only 3 
          # elements, so the last is item 2

IndexError: list index out of range

In [55]:
mylist[-1] # negative index starts from the last element

15

In [56]:
mylist[1:] # : is called slice, all items starting from the second 
           # (remember indexing starts from 0)

[74, 15]

In [57]:
mylist[:2] # elements up to but excluding item 2

[21, 74]

In [3]:
# the following two loops are equivalent
for i in range(len(mylist)): 
    print (mylist[i])
    
for l in mylist:
    print (l)

NameError: name 'mylist' is not defined

In [61]:
for i, item in enumerate(mylist): # enumerate keyword iterates
                                  # returning the item and its index
    print (i, item)

0 21
1 74
2 85
3 15
4 188


In [58]:
mylist.append(188) # append add an item at the end of the list
mylist

[21, 74, 15, 188]

In [59]:
mylist.insert(2, 85) # insert an item in the desired position 
                     # (2 in this example)
mylist

[21, 74, 85, 15, 188]

In [71]:
# lists can be made of different kind of objects
mixedlist = [1, 2, "b", math.sqrt]
print (mixedlist)

[1, 2, 'b', <built-in function sqrt>]


### Dictionaries

Dictionaries are objects which map keys to values. Keys can be (almost) any kind of object (strings, numbers...) and they do not have to be sequential.

$$"apple" \rightarrow 4 $$
$$"banana" \rightarrow 5 $$

Dictionaries are very flexible so that different keys can be different type of objects.

In [73]:
adict = {"apple": 4, "banana": 5}
adict["apple"] # values are accessed through their key

4

In [74]:
adict["pear"] # error ! this key doesn't exists

KeyError: 'pear'

In [75]:
"pear" in adict # indeed

False

In [76]:
len(adict)

2

In [77]:
adict["banana"] = 2
print (len(adict))
print (adict)

2
{'apple': 4, 'banana': 2}


In [78]:
adict[math.log] = math.exp

In [79]:
adict.keys() # returns all the keys

dict_keys(['apple', 'banana', <built-in function log>])

In [80]:
for key in adict.keys():
    print (key)

apple
banana
<built-in function log>


In [81]:
for value in adict.values():
    print (value)

4
2
<built-in function exp>


In [82]:
for item in adict.items(): # items() returns a pair of key, value
    print (item)

('apple', 4)
('banana', 2)
(<built-in function log>, <built-in function exp>)


In [83]:
seconddict = {"watermelon": 0, "strawberry": 1}
adict.update(seconddict) # update adict with the items in seconddict
adict

{'apple': 4,
 'banana': 2,
 <function math.log>: <function math.exp>,
 'watermelon': 0,
 'strawberry': 1}

## Exercises

### Exercise 1.1

* What is the built-in function that Python uses to iterate over a number sequence?
* What is a string in Python?
* What does the continue do in Python?
* What is the purpose of end in Python?
* When should you use the break in Python?
* What is a dictionary in Python programming?
* What is the use of the dictionary in Python?
* How do you create a dictionary in Python?
* How do you read from a dictionary in Python?
* How do you traverse through a dictionary object in Python?
* How do you add elements To a dictionary in Python?
* How do you delete elements Of a dictionary in Python?
* How do you check the presence of a key in a dictionary?
* What is the syntax for list comprehension in Python?
* What is the syntax for dictionary comprehension in Python?
* How do you write a conditional expression in Python?
* Which Python function will you use to convert a number to a string?


# Advanced hints

If you would like to know much **more** about python during the course without the EDX site you can look these videos during the course in your spare time:

Some more information about different kind of languages:
https://www.youtube.com/watch?v=9oYFH4OmYDY

Basic data types:
https://www.youtube.com/watch?v=XIjrEt2lz1U

Variables:
https://www.youtube.com/watch?v=z2NLjdfxEyQ

Branching:
https://www.youtube.com/watch?v=8vr3nyg5QcM

Tuples:
https://www.youtube.com/watch?v=CwZyWaap5Z8

Lists:
https://www.youtube.com/watch?v=eMyWO0tcxKg

List Operations:
https://www.youtube.com/watch?v=rQBho4-bI3o

Mutation, Aliasing, Cloning:
https://www.youtube.com/watch?v=2SRXg8Or-Pc

Functions as Objects:
https://www.youtube.com/watch?v=pheM3rVmGMU

Dictionaries:
https://www.youtube.com/watch?v=elSt5hke-Rs